src/main/java/com/hms_networks/americas/sc/thingworx/TWConnectorMain.java

```diff
@@ -121,114 +125,193 @@ private static void runGrabData() {
+
+                // Check if a queue force reset has been requested
+                boolean forceReset = false;
+                if (queueDiagnosticForceResetTag != null) {
+                  forceReset =
+                      queueDiagnosticForceResetTag.getTagValueAsInt()
+                          == TWConnectorConsts.QUEUE_DIAGNOSTIC_TAG_FORCE_RESET_TRUE_VALUE;
+
+                  if (forceReset) {
+                    Logger.LOG_SERIOUS(
+                        "A force reset of the queue has been requested using the tag `"
+                            + TWConnectorConsts.QUEUE_DIAGNOSTIC_TAG_FORCE_RESET_NAME
+                            + "`. A new queue time tracker will be created at the current time!");
+                  }
+                }
+
+                // Read data points from queue
+                ArrayList datapointsReadFromQueue;
+                if (HistoricalDataQueueManager.doesTimeTrackerExist() && !forceReset) {
+                  final boolean startNewTimeTracker = false;
+                  datapointsReadFromQueue =
+                      HistoricalDataQueueManager.getFifoNextSpanDataAllGroups(startNewTimeTracker);
+                } else {
+                  final boolean startNewTimeTracker = true;
+                  datapointsReadFromQueue =
+                      HistoricalDataQueueManager.getFifoNextSpanDataAllGroups(startNewTimeTracker);
+                }
+
+                // Reset queue force reset tag value to false, if true
+                if (queueDiagnosticForceResetTag != null && forceReset) {
+                  queueDiagnosticForceResetTag.setTagValueAsInt(
+                      TWConnectorConsts.QUEUE_DIAGNOSTIC_TAG_FORCE_RESET_FALSE_VALUE);
+                }
+
+                Logger.LOG_DEBUG(
+                    "Read "
+                        + datapointsReadFromQueue.size()
+                        + " data points from the historical log.");
+
+                // Send data to Thingworx
+                TWDataManager.addDataPointsToPending(datapointsReadFromQueue);
+
+                // Check if queue is behind
+                try {
+                  long queueBehindMillis = HistoricalDataQueueManager.getQueueTimeBehindMillis();
+                  if (queueBehindMillis >= TWConnectorConsts.QUEUE_DATA_POLL_BEHIND_MILLIS_WARN) {
+                    String timeBehindString =
+                        SCTimeUtils.getDayHourMinSecsForMillis(
+                            (int) queueBehindMillis,
+                            DAYS_STRING,
+                            HOURS_STRING,
+                            MINUTES_STRING,
+                            SECONDS_STRING);
+                    Logger.LOG_WARN(
+                        "The historical data queue is running behind by " + timeBehindString + ".");
+                  } else {
+                    // Queue is not past running behind threshold, reset to 0 for updating debug tag
+                    queueBehindMillis = 0;
```

Comment:
0 could be TWConnectorConsts.QUEUE_DATA_POLL_CURRENT_MILLIS or something similar.